### PR TITLE
Add Codespaces support

### PR DESCRIPTION
**Pull request by potiuk**
Codespaces is (soon to be GA) feature that allows to start
development environment on a virtual machine directly from the
GitHub UI or from VSCode connected with GitHub account.

This PR adds configuration that allows to start codespaces
environment (based on Breeze's docker compose environment)
in the basic form that allows the users to run basic unit
tests.

It does not replace Breeze/CI but for most users who want
to just run a few tests, make sure that all dependencies
and plugins are installed so that the users can get the
VSCode autocomplete and execution of the tests possible.

This is a great way to begin your journey with Airflow - while
the Codespace machines are not very powerful and the integration
tests might require more resources, the "no-setup"
environment and no need to have powerful machine is great
for contributors who cannot afford powerful development
workstations.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
This is a very basic suppor for codespaces, but it works nicely.

For those who have access to codespaces (it's not yet GA for Apache and Public repos) it takes about 4 minutes to get you up and running:

You land inside "breeze" container with terminal and pytest working, postgres database and you can run unit tests from within VSCode (either in-browser or local). 

It required to move the helm charts to "tests/charts folder" because of the way how VSCode does NOT handle well mutliple folders with pytest conftest.py in separate "test trees".

This is a small inconvenience (if at all) but I wanted you to take  a look @jedcunningham @kaxil @ephraimbuddy.

Also as a side effect of it you can run just `pytest` in the root folder of airflow and all tests (including the helm chart tests) will get executed (and conftest.py will be properly applied to the subset of tests in each subfolder). This brings our test structure more in line with the "standard" pytest test tree approach.

**Comment by potiuk** [comment]
I think it is totally worth it - especially for new contributors. My goal of < 10 minutes to get Airflow environment up-and-running is well exceeded with that one.

**Comment by potiuk** [comment]
Need to work a bit on executing the tests on CI :)

**Comment by potiuk** [comment]
I hope it will be better. Also @jedcunningham @kaxil @ephraimbuddy @ashb -> you might get happy with that  change, as I also enabled parallel tests for the Helm Chart tests in this change.

It will immensely speed up the Helm chart Job I thinlk. Even on the super-fast machines we have in Self-Hosted runs it took 25 minutes to run. On my machine when I enabled parallel execution it took < 8 minutes so we might have a huge, huge, huge speedup. :crossed_fingers: 

image

**Comment by potiuk** [comment]
Ah no.... they were already enabled as auto. I am not sure why they take 3 times more  in CI.

**Review by dstandish — APPROVED** [review]
pretty cool 👍

**Comment by potiuk** [comment]
Update: Setting up a working test-ready Codespace for airlfow takes ~ 3 minutes.

**Comment by potiuk** [comment]
> pretty cool +1

It is indeed :)

**Comment by dstandish** [comment]
> Might make me switch to VSCode actually.

you a pycharm user?  i've tried it out now and then... always run back to pycharm....  but will have to try this

**Comment by potiuk** [comment]
> you a pycharm user? i've tried it out now and then... always run back to pycharm.... but will have to try this

Intellij Ultimate for many years. But recent event made me rethink where my money go, so even if this is my main productivity tool for years (except `vi` of course which is probably more so), VSCode is great alternative.

**Comment by potiuk** [comment]
Right. I got it green now.

Please @jedcunningham @ephraimbuddy @kaxil  take a look at the move of tests and whether the move of tests is "OK" for you,

The added benefit of this move is that integration with any other IDE for tests is now a BREEZE :). Tests will be automatically recognized and configured, without any special setup or special projects (i actually had to have previously multiple projects  in my IDE to be able to run Helm tests - this is not needed any more. 

Should be much easier to start their journey with Airflow now :)

BTW. I am not sure if you have codespaces or not... But if you do ....  Simply press "." on this PR to try it out.

**Comment by potiuk** [comment]
Ah no. Needs to be merged to `main` in order to apply for `apache/airflow` repo :).

**Comment by dstandish** [comment]
> > pretty cool +1
> 
> BTW. @dstandish For me this is a culmination of 3.5 years of devenv work.
> 
> All the Breeze, constraint, ci docker image etc. All of it was there to eventually enable this one seamlessly (I knew it was coming). This builds on top of all the automation that is behind maintaining our CI images and constraints updated so that whenever you start codespace on a new PR, you will get the latest and greatest environment to work on in < 10 minutes (we got down to < 3).
> 
> This is beyond cool (but let's see how it is going to be useful - it is still not GA for all contributors unfortunately. It's only available as GA for Teams and GitHub Enterprise Cloud customers (and those users like me who applied for Beta access). When and how it will be available for personal repos/Open Source repos - we do not know (but I am meeting my friend in Berlin next week who works there and I hope I will know more :) .

yes, "pretty cool" does feel quite inadequate in hindsight :)
it's quite a mountain that you have climbed. i remember the days of the puckel image and when breeze was still being incubated.

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -197,15 +197,24 @@ From the `apache/airflow <https://github.com/apache/airflow>`_ repo,
 Step 2: Configure Your Environment
 ----------------------------------
 
-You can use either a local virtual env or a Docker-based env. The differences
-between the two are explained `here <https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#development-environments>`__.
+You can use use several development environments for Airflow. If you prefer to have development environments
```
Comment (ephraimbuddy) [inline CONTRIBUTING.rst]:
```suggestion
You can use several development environments for Airflow. If you prefer to have development environments
```

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -472,25 +481,27 @@ develop Apache Airflow:
...
-------------------------- -------------------------------- -------------------------------------
-Reproducing CI failures   - (-) not possible in many cases - (+) fully reproducible
-------------------------- -------------------------------- -------------------------------------
-Ability to update         - (-) requires manual updates    - (+) automated update via breeze cmd
-------------------------- -------------------------------- -------------------------------------
-Disk space and CPU usage  - (+) relatively lightweight     - (-) uses GBs of disk and many CPUs
-------------------------- -------------------------------- -------------------------------------
-IDE integration           - (+) straightforward            - (-) via remote debugging only
-========================= ================================ =====================================
+========================= ================================ ===================================== ========================================
```
Comment (ephraimbuddy) [inline CONTRIBUTING.rst:484]: The line at 481 ` two environments:` can be changed to `three environments`

File: scripts/ci/libraries/_build_images.sh [hunk 3]
```diff
@@ -598,7 +598,7 @@ function build_images::build_prod_images() {
             "--cache-from=${AIRFLOW_PROD_IMAGE}:cache"
         )
     else
-        echo
+        echod
```
Comment (ephraimbuddy) [inline scripts/ci/libraries/_build_images.sh]:
```suggestion
        echo
```
Typo?

**Review by ephraimbuddy — APPROVED** [review]
LGTM

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -472,25 +481,27 @@ develop Apache Airflow:
 The table below summarizes differences between the two environments:
```
Comment (potiuk) [inline CONTRIBUTING.rst]:
```suggestion
The table below summarizes differences between the environments:
```

**Comment by potiuk** [comment]
Merged!

**Comment by kaxil** [comment]
Nice, well done. I haven't looked at the exact code changes but reading by the PR description, this looks great.

**Pull request merged**